<div class="Markdown-module__Markdown--f5ALc TasksBankTaskCard-module__TaskCardContentMarkdown--hsXqP"><p>Астроном решил провести кластеризацию некоторого множества звёзд по их расположению на карте звёздного неба. Область наблюдения представляет собой круг с центром в точке (0, 0). Кластер звёзд — это набор звёзд (точек) на графике, который лежит внутри сектора этого круга с определённым углом <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi></mrow><annotation encoding="application/x-tex">\alpha</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.4306em;"></span><span class="mord mathnormal" style="margin-right: 0.0037em;">α</span></span></span></span></span>. Каждая звезда обязательно принадлежит только одному из кластеров.</p>
<p>Истинный центр кластера, или <strong>центроид</strong>, — это одна из звёзд на графике, сумма расстояний от которой до всех остальных звёзд кластера минимальна.</p>
<p>Под расстоянием понимается расстояние Евклида между двумя точками <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi><mo stretchy="false">(</mo><msub><mi>x</mi><mn>1</mn></msub><mo separator="true">,</mo><msub><mi>y</mi><mn>1</mn></msub><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">A(x_1, y_1)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mord mathnormal">A</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: 0em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: -0.0359em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mclose">)</span></span></span></span></span> и <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>B</mi><mo stretchy="false">(</mo><msub><mi>x</mi><mn>2</mn></msub><mo separator="true">,</mo><msub><mi>y</mi><mn>2</mn></msub><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">B(x_2, y_2)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mord mathnormal" style="margin-right: 0.05017em;">B</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: 0em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: -0.0359em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mclose">)</span></span></span></span></span> на плоскости, которое вычисляется по формуле:</p>
<p><span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>d</mi><mo stretchy="false">(</mo><mi>A</mi><mo separator="true">,</mo><mi>B</mi><mo stretchy="false">)</mo><mo>=</mo><msqrt><mrow><mo stretchy="false">(</mo><msub><mi>x</mi><mn>2</mn></msub><mo>−</mo><msub><mi>x</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup><mo>+</mo><mo stretchy="false">(</mo><msub><mi>y</mi><mn>2</mn></msub><mo>−</mo><msub><mi>y</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></msqrt><mi mathvariant="normal">.</mi></mrow><annotation encoding="application/x-tex">d(A, B) = \sqrt{(x_2 - x_1)^2 + (y_2 - y_1)^2}.</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord mathnormal">A</span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord mathnormal" style="margin-right: 0.05017em;">B</span><span class="mclose">)</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 1.24em; vertical-align: -0.305em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.935em;"><span class="svg-align" style="top: -3.2em;"><span class="pstrut" style="height: 3.2em;"></span><span class="mord" style="padding-left: 1em;"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: 0em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: 0em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height: 0.7401em;"><span style="top: -2.989em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: -0.0359em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: -0.0359em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height: 0.7401em;"><span style="top: -2.989em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span><span style="top: -2.895em;"><span class="pstrut" style="height: 3.2em;"></span><span class="hide-tail" style="min-width: 1.02em; height: 1.28em;"><svg xmlns="http://www.w3.org/2000/svg" width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.305em;"><span></span></span></span></span></span><span class="mord">.</span></span></span></span></span></p>
<p>В файле A хранятся данные о звёздах <strong>двух</strong> кластеров. В&nbsp;каждой строке записана информация о расположении на&nbsp;карте одной звезды в&nbsp;полярной системе координант: сначала полярный угол <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ϕ</mi></mrow><annotation encoding="application/x-tex">\phi</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.8889em; vertical-align: -0.1944em;"></span><span class="mord mathnormal">ϕ</span></span></span></span></span> в&nbsp;градусах, затем полярный радиус <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>r</mi></mrow><annotation encoding="application/x-tex">r</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.4306em;"></span><span class="mord mathnormal" style="margin-right: 0.02778em;">r</span></span></span></span></span> в&nbsp;условных единицах. Известно, что количество звёзд не&nbsp;превышает 1000.</p>
<p>В файле&nbsp;Б&nbsp;хранятся данные о&nbsp;звёздах <strong>трёх</strong> кластеров. Известно, что количество звёзд не&nbsp;превышает 10&nbsp;000. Структура хранения информации о&nbsp;звездах в&nbsp;файле Б&nbsp;аналогична файлу&nbsp;А.</p>
<p>Для каждого файла определите координаты центра каждого кластера, затем вычислите два числа: <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>P</mi><mi>x</mi></msub></mrow><annotation encoding="application/x-tex">P_x</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.8333em; vertical-align: -0.15em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.1514em;"><span style="top: -2.55em; margin-left: -0.1389em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">x</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span></span></span></span></span>&nbsp;— среднее арифметическое абсцисс центров кластеров и&nbsp;<span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>P</mi><mi>y</mi></msub></mrow><annotation encoding="application/x-tex">P_y</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.9694em; vertical-align: -0.2861em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.1514em;"><span style="top: -2.55em; margin-left: -0.1389em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right: 0.03588em;">y</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.2861em;"><span></span></span></span></span></span></span></span></span></span></span>&nbsp;— среднее арифметическое ординат центров кластеров.</p>
<p>В&nbsp;ответе запишите четыре числа: в&nbsp;первой строке сначала целую часть произведения <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>P</mi><mi>x</mi></msub><mo>×</mo><mn>10</mn><mtext>&thinsp;</mtext><mn>000</mn></mrow><annotation encoding="application/x-tex">P_x \times 10\, 000</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.8333em; vertical-align: -0.15em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.1514em;"><span style="top: -2.55em; margin-left: -0.1389em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">x</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">×</span><span class="mspace" style="margin-right: 0.2222em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">10</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord">000</span></span></span></span></span>, затем целую часть произведения <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>P</mi><mi>y</mi></msub><mo>×</mo><mn>10</mn><mtext>&thinsp;</mtext><mn>000</mn></mrow><annotation encoding="application/x-tex">P_y \times 10\,000</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.9694em; vertical-align: -0.2861em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.1514em;"><span style="top: -2.55em; margin-left: -0.1389em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right: 0.03588em;">y</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.2861em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">×</span><span class="mspace" style="margin-right: 0.2222em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">10</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord">000</span></span></span></span></span> для файла A, во второй строке — аналогичные данные для файла Б.</p>
<p>Возможные данные одного из файлов иллюстрированы графиком.</p>
<p><img src="https://avatars.mds.yandex.net/get-schoolbook-images/5519467/471faafa35aa9c30c3c98cedd677a610/orig" alt="Screenshot"></p>
<p><strong>Справка о полярной системе координат:</strong></p>
<p>Полярная система координат — это двумерная система координат, в которой каждая точка на плоскости определяется двумя числами: полярным углом и полярным радиусом.</p>
<ul>
<li>Полярная ось — это фиксированная прямая, от которой отсчитывается полярный угол. Важно отметить, что полярная ось совпадает с положительным направлением оси абсцисс (оси <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>X</mi></mrow><annotation encoding="application/x-tex">X</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.6833em;"></span><span class="mord mathnormal" style="margin-right: 0.07847em;">X</span></span></span></span></span>) в декартовой системе координат.</li>
<li>Полярный угол (<span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ϕ</mi></mrow><annotation encoding="application/x-tex">\phi</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.8889em; vertical-align: -0.1944em;"></span><span class="mord mathnormal">ϕ</span></span></span></span></span>) — это угол между полярной осью и линией, соединяющей точку с началом координат. Угол отсчитывается против часовой стрелки от полярной оси.</li>
<li>Полярный радиус (<span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>r</mi></mrow><annotation encoding="application/x-tex">r</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.4306em;"></span><span class="mord mathnormal" style="margin-right: 0.02778em;">r</span></span></span></span></span>) — это расстояние от точки до начала координат.</li>
</ul>
<p>Переход из полярной системы координат в декартову:</p>
<p><span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>x</mi><mo>=</mo><mi>r</mi><mo>⋅</mo><mi>c</mi><mi>o</mi><mi>s</mi><mo stretchy="false">(</mo><mi>ϕ</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">x = r \cdot cos(\phi)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.4306em;"></span><span class="mord mathnormal">x</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.4445em;"></span><span class="mord mathnormal" style="margin-right: 0.02778em;">r</span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right: 0.2222em;"></span></span><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mord mathnormal">cos</span><span class="mopen">(</span><span class="mord mathnormal">ϕ</span><span class="mclose">)</span></span></span></span></span></p>
<p><span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>y</mi><mo>=</mo><mi>r</mi><mo>⋅</mo><mi>s</mi><mi>i</mi><mi>n</mi><mo stretchy="false">(</mo><mi>ϕ</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">y = r \cdot sin(\phi)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.625em; vertical-align: -0.1944em;"></span><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.4445em;"></span><span class="mord mathnormal" style="margin-right: 0.02778em;">r</span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right: 0.2222em;"></span></span><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mord mathnormal">s</span><span class="mord mathnormal">in</span><span class="mopen">(</span><span class="mord mathnormal">ϕ</span><span class="mclose">)</span></span></span></span></span></p>
<p>где <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>x</mi></mrow><annotation encoding="application/x-tex">x</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.4306em;"></span><span class="mord mathnormal">x</span></span></span></span></span> и <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>y</mi></mrow><annotation encoding="application/x-tex">y</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.625em; vertical-align: -0.1944em;"></span><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span></span></span></span></span> — координаты в декартовой системе, <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>r</mi></mrow><annotation encoding="application/x-tex">r</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.4306em;"></span><span class="mord mathnormal" style="margin-right: 0.02778em;">r</span></span></span></span></span> — полярный радиус, <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ϕ</mi></mrow><annotation encoding="application/x-tex">\phi</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.8889em; vertical-align: -0.1944em;"></span><span class="mord mathnormal">ϕ</span></span></span></span></span> — полярный угол в градусах.</p>
<p>Поскольку полярная ось совпадает с осью абсцисс, точка с полярными координатами <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="false">(</mo><msup><mn>0</mn><mo lspace="0em" rspace="0em">∘</mo></msup><mo separator="true">,</mo><mi>r</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">(0^{\circ}, r)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mopen">(</span><span class="mord"><span class="mord">0</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height: 0.6741em;"><span style="top: -3.063em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">∘</span></span></span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord mathnormal" style="margin-right: 0.02778em;">r</span><span class="mclose">)</span></span></span></span></span> будет иметь декартовы координаты <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="false">(</mo><mn>0</mn><mo separator="true">,</mo><mi>r</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">(0, r)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mopen">(</span><span class="mord">0</span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord mathnormal" style="margin-right: 0.02778em;">r</span><span class="mclose">)</span></span></span></span></span>, а точка <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="false">(</mo><mn>9</mn><msup><mn>0</mn><mo lspace="0em" rspace="0em">∘</mo></msup><mo separator="true">,</mo><mi>r</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">(90^{\circ}, r)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mopen">(</span><span class="mord">9</span><span class="mord"><span class="mord">0</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height: 0.6741em;"><span style="top: -3.063em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">∘</span></span></span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord mathnormal" style="margin-right: 0.02778em;">r</span><span class="mclose">)</span></span></span></span></span> — координаты <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="false">(</mo><mi>r</mi><mo separator="true">,</mo><mn>0</mn><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">(r, 0)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right: 0.02778em;">r</span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord">0</span><span class="mclose">)</span></span></span></span></span>.</p></div>

Ответ:

-182448 -202245

10106 528042

In [3]:
import math

a = []

with open("./1/27_A (1).txt") as f:
    for i in f.readlines():
        a.append(list(map(float, i.split())))
        
for i in range(len(a)):
    a[i] = a[i][1] * math.cos(math.radians(a[i][0])), a[i][1] * math.sin(math.radians(a[i][0]))
    

clust = [[], []]

for i in a:
    clust_num = 0 if i[0] < 0 else 1
    clust[clust_num].append(i)

def dist(a, b):
    return math.sqrt((a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2)

def sum_dist(a, clust):
    return sum(dist(a, j) for j in clust)

ans = []

for j in range(len(clust)):
    mini = 1e9
    v = -1
    for i in range(len(clust[j])):
        val = sum_dist(clust[j][i], clust[j])
        if val < mini:
            mini = val
            v = clust[j][i]
    ans.append(v)
print(sum(i[0] for i in ans) / 2, sum(i[1] for i in ans) / 2)

-18.244821256520854 -20.22450566486108


In [ ]:
import math

a = []

with open("./1/27_B (1).txt") as f:
    for i in f.readlines():
        a.append(list(map(float, i.split())))
        
for i in range(len(a)):
    a[i] = a[i][1] * math.cos(math.radians(a[i][0])), a[i][1] * math.sin(math.radians(a[i][0]))
    

clust = [[], [], []]

for i in a:
    clust_num = 0 if i[0] < 0 else 1  # TODO
    clust[clust_num].append(i)

def dist(a, b):
    return math.sqrt((a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2)

def sum_dist(a, clust):
    return sum(dist(a, j) for j in clust)

ans = []

for j in range(len(clust)):
    mini = 1e9
    v = -1
    for i in range(len(clust[j])):
        val = sum_dist(clust[j][i], clust[j])
        if val < mini:
            mini = val
            v = clust[j][i]
    ans.append(v)
print(sum(i[0] for i in ans) / 2, sum(i[1] for i in ans) / 2)

<div class="Text Text_typography_bodyM TasksBankTaskCard-module__TaskCardContent--PNxia"><div class="Markdown-module__Markdown--f5ALc TasksBankTaskCard-module__TaskCardContentMarkdown--hsXqP"><p>Лесник решил провести кластеризацию животных по&nbsp;их&nbsp;расположению в&nbsp;лесу. Он&nbsp;разделил карту на&nbsp;квадраты по&nbsp;километровым отметкам: первый квадрат можно описать <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>0</mn><mtext>&nbsp;</mtext><mo>≤</mo><mtext>&nbsp;</mtext><mi>x</mi><mtext>&nbsp;</mtext><mo>≤</mo><mtext>&nbsp;</mtext><mn>1</mn><mo separator="true">,</mo><mtext>&nbsp;</mtext><mn>0</mn><mtext>&nbsp;</mtext><mo>≤</mo><mtext>&nbsp;</mtext><mi>y</mi><mtext>&nbsp;</mtext><mo>≤</mo><mtext>&nbsp;</mtext><mn>1</mn></mrow><annotation encoding="application/x-tex">0~\leq~x~\leq~1,~0~\leq~y~\leq~1</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.8389em; vertical-align: -0.1944em;"></span><span class="mord">0</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mord mathnormal">x</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mord">1</span><span class="mpunct">,</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord">0</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mord">1</span></span></span></span></span>, второй&nbsp;— <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mtext>&nbsp;</mtext><mo>≤</mo><mtext>&nbsp;</mtext><mi>x</mi><mtext>&nbsp;</mtext><mo>≤</mo><mtext>&nbsp;</mtext><mn>2</mn><mo separator="true">,</mo><mn>0</mn><mtext>&nbsp;</mtext><mo>≤</mo><mtext>&nbsp;</mtext><mi>y</mi><mtext>&nbsp;</mtext><mo>≤</mo><mtext>&nbsp;</mtext><mn>1</mn></mrow><annotation encoding="application/x-tex">1~\leq~x~\leq~2, 0~\leq~y~\leq~1</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.8389em; vertical-align: -0.1944em;"></span><span class="mord">1</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mord mathnormal">x</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mord">2</span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord">0</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace nobreak">&nbsp;</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mord">1</span></span></span></span></span> и&nbsp;так далее.</p>
<p>Для файла А&nbsp;леснику нужно определить два соседних квадрата, в&nbsp;которых суммарно находится больше всего животных. Для файла Б&nbsp;леснику нужно определить три соседних квадрата. Квадраты называются соседними, если у&nbsp;них есть общая граница.</p>
<p>Для каждого файла вычислите два числа: S&nbsp;— количество социальных животных в&nbsp;выбранных соседних квадратах, и&nbsp;K&nbsp;— количество остальных животных в&nbsp;выбранных соседних квадратах. Животное называется социальным, если в&nbsp;радиусе&nbsp;0.1 вокруг него находится как&nbsp;минимум 14&nbsp;других животных.</p>
<p>В&nbsp;ответе запишите четыре числа: в&nbsp;первой строке S&nbsp;и&nbsp;K&nbsp;для файла&nbsp;А, во&nbsp;второй строке аналогичные данные для файла&nbsp;Б.</p>
<p>Возможные данные одного из&nbsp;файлов иллюстрированы графиком.</p>
<p><strong>Внимание! График приведён в&nbsp;иллюстративных целях для произвольных значений и&nbsp;не&nbsp;имеет отношения к&nbsp;заданию.</strong></p>
<p><strong>Чтобы выполнить задание, используйте данные из&nbsp;файла в&nbsp;приложении.</strong></p>
<p><img src="https://avatars.mds.yandex.net/get-schoolbook-images/6887102/3b494a4eb25ddb0131f2fde1d4366053/orig" alt="Screenshot"></p></div></div>

Ответ:

104 453

2156 158

In [4]:
import math

a = []

with open("./2/27_A (2).txt") as f:
    for i in f.readlines():
        a.append(list(map(float, i.split())))
        
clust = [[[] for i in range(3)] for j in range(3)]

for i in a:
    clust[int(i[0])][int(i[1])].append(i)
    
for i in clust:
    for j in i:
        print(len(j), end=' ')
    print()

55 72 68 
69 288 68 
81 269 80 


In [6]:
maxi = 0
idx = []

for i in range(9):
    for j in range(i + 1, 9):
        x1, y1, x2, y2 = i // 3, i % 3, j // 3, j % 3
        if (x1 == x2 and abs(y1 - y2) == 1) or (abs(x1 - x2) == 1 and y1 == y2):
            if maxi < len(clust[x1][y1]) + len(clust[x2][y2]):
                maxi = len(clust[x1][y1]) + len(clust[x2][y2])
                idx = [x1, y1, x2, y2]
            # print(x1, y1, x2, y2)

sosial = 0

for i in clust[idx[0]][idx[1]]:
    cnt = -1
    for j in a:
        if math.sqrt((j[0] - i[0]) ** 2 + (j[1] - i[1]) ** 2) <= 0.1:
            cnt += 1
    if cnt >= 14:
        sosial += 1
        
for i in clust[idx[2]][idx[3]]:
    cnt = -1
    for j in a:
        if math.sqrt((j[0] - i[0]) ** 2 + (j[1] - i[1]) ** 2) <= 0.1:
            cnt += 1
    if cnt >= 14:
        sosial += 1
        
print(sosial, maxi - sosial)

104 453


In [7]:
import math

a = []

with open("./2/27_B (2).txt") as f:
    for i in f.readlines():
        a.append(list(map(float, i.split())))
        
clust = [[[] for i in range(5)] for j in range(5)]

for i in a:
    clust[int(i[0])][int(i[1])].append(i)
    
for i in clust:
    for j in i:
        print(len(j), end=' ')
    print()
    
# делаем руками
idx = [2, 3, 3, 3, 3, 2]
maxi = len(clust[idx[0]][idx[1]]) + len(clust[idx[2]][idx[3]]) + len(clust[idx[4]][idx[5]])
#
sosial = 0

for i in clust[idx[0]][idx[1]]:
    cnt = -1
    for j in a:
        if math.sqrt((j[0] - i[0]) ** 2 + (j[1] - i[1]) ** 2) <= 0.1:
            cnt += 1
    if cnt >= 14:
        sosial += 1
        
for i in clust[idx[2]][idx[3]]:
    cnt = -1
    for j in a:
        if math.sqrt((j[0] - i[0]) ** 2 + (j[1] - i[1]) ** 2) <= 0.1:
            cnt += 1
    if cnt >= 14:
        sosial += 1
        
for i in clust[idx[4]][idx[5]]:
    cnt = -1
    for j in a:
        if math.sqrt((j[0] - i[0]) ** 2 + (j[1] - i[1]) ** 2) <= 0.1:
            cnt += 1
    if cnt >= 14:
        sosial += 1
print(sosial, maxi - sosial)


188 210 196 179 217 
189 208 183 230 198 
214 204 226 825 211 
195 195 762 727 203 
210 173 217 210 179 
2156 158


<div class="Text Text_typography_bodyM TasksBankTaskCard-module__TaskCardContent--PNxia"><div class="Markdown-module__Markdown--f5ALc TasksBankTaskCard-module__TaskCardContentMarkdown--hsXqP"><p>HR-менеджер решил провести анализ вакансий на&nbsp;рынке труда. Он&nbsp;попросил отдел разработки собрать информацию из&nbsp;доступных интернет-источников и&nbsp;получил от&nbsp;них два текстовых файла, в&nbsp;которых значения разделены точкой с&nbsp;запятой.</p>
<p>Первая строка каждого файла содержит имена полей:
ID; Название_вакансии; Компания; Город; Мин_зарплата; Макс_зарплата; Опыт_работы; Дата_публикации; Требуемые_навыки</p>
<p>Пример строки файла:<br>
<code>1001;Python Developer;TechCorp;Москва;120000;200000;3-5 лет;2024-03-15;[Python,Django,SQL]</code></p>
<p>HR-менеджер разделил вакансии на группы по&nbsp;уровню средней зарплаты и&nbsp;опыту работы. Уровни зарплаты были заданы с&nbsp;шагом 50000&nbsp;рублей: <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>0</mn><mo>≤</mo><mtext>зарплата</mtext><mo>≤</mo><mn>50000</mn></mrow><annotation encoding="application/x-tex">0 \le \text {зарплата} \le 50000</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.7804em; vertical-align: -0.136em;"></span><span class="mord">0</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.8304em; vertical-align: -0.1944em;"></span><span class="mord text"><span class="mord cyrillic_fallback">зарплата</span></span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">50000</span></span></span></span></span>, <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>50000</mn><mo>≤</mo><mtext>зарплата</mtext><mo>≤</mo><mn>100000</mn></mrow><annotation encoding="application/x-tex">50000 \le \text {зарплата} \le 100000</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.7804em; vertical-align: -0.136em;"></span><span class="mord">50000</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.8304em; vertical-align: -0.1944em;"></span><span class="mord text"><span class="mord cyrillic_fallback">зарплата</span></span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">100000</span></span></span></span></span> и так далее. Опыт работы разбит на категории: 0-1, 2-3, 4-5, 6-7, 8-9 лет. Примеры групп: первая — <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>0</mn><mo>≤</mo><mtext>зарплата</mtext><mo>≤</mo><mn>50000</mn></mrow><annotation encoding="application/x-tex">0 \le \text {зарплата} \le 50000</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.7804em; vertical-align: -0.136em;"></span><span class="mord">0</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.8304em; vertical-align: -0.1944em;"></span><span class="mord text"><span class="mord cyrillic_fallback">зарплата</span></span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">50000</span></span></span></span></span>, <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>0</mn><mo>≤</mo><mtext>опыт</mtext><mo>≤</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">0 \le \text {опыт} \le 1</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.7804em; vertical-align: -0.136em;"></span><span class="mord">0</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.7719em; vertical-align: -0.136em;"></span><span class="mord text"><span class="mord cyrillic_fallback">опыт</span></span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">1</span></span></span></span></span>, вторая — <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>50000</mn><mo>≤</mo><mtext>зарплата</mtext><mo>≤</mo><mn>100000</mn></mrow><annotation encoding="application/x-tex">50000 \le \text {зарплата} \le 100000</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.7804em; vertical-align: -0.136em;"></span><span class="mord">50000</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.8304em; vertical-align: -0.1944em;"></span><span class="mord text"><span class="mord cyrillic_fallback">зарплата</span></span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">100000</span></span></span></span></span>, <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>2</mn><mo>≤</mo><mtext>опыт</mtext><mo>≤</mo><mn>3</mn></mrow><annotation encoding="application/x-tex">2 \le \text {опыт} \le 3</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.7804em; vertical-align: -0.136em;"></span><span class="mord">2</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.7719em; vertical-align: -0.136em;"></span><span class="mord text"><span class="mord cyrillic_fallback">опыт</span></span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">≤</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">3</span></span></span></span></span>.</p>
<p>Чтобы получить точный результат, HR-менеджер решил не&nbsp;учитывать строки, в&nbsp;которых:</p>
<ol>
<li>пропущены значения в&nbsp;любых полях, кроме полей минимальной и&nbsp;максимальной зарплаты</li>
<li>минимальная зарплата указана больше максимальной</li>
</ol>
<p>Также он&nbsp;обнаружил, что в&nbsp;файлах есть дублирующиеся записи и&nbsp;оставил только первые вхождения, а&nbsp;повторы удалил.</p>
<p>Для группировки используется понятие средней зарплаты, которая вычисляется как среднее арифметическое минимальной и&nbsp;максимальной. Если одно из&nbsp;этих значений отсутствует, за&nbsp;среднюю зарплату принимается известное значение.</p>
<p>HR-менеджер решил провести анализ в&nbsp;соседних группах вакансий. Группы называются соседними, если у&nbsp;них есть общая граница по&nbsp;зарплате или опыт работы попадает в&nbsp;соседние категории (например, 0-1 и&nbsp;2-3 или 4-5 и&nbsp;6-7). Для файла A&nbsp;HR-менеджеру нужно определить две соседние группы вакансий, в&nbsp;которых суммарно находится больше всего вакансий. Для файла B&nbsp;ему нужно определить три группы, которые будут попарно соседними.</p>
<p>Для каждого файла вычислите два числа: S&nbsp;— количество уникальных вакансий в&nbsp;выбранных соседних группах, для которых требуется знание Python, и&nbsp;K&nbsp;— количество вакансий в&nbsp;выбранных соседних группах, для которых знание Python не&nbsp;требуется.</p>
<p>В&nbsp;ответе запишите четыре числа: в&nbsp;первой строке S&nbsp;и&nbsp;K для&nbsp;файла&nbsp;A, во&nbsp;второй строке аналогичные данные для&nbsp;файла&nbsp;B.</p>
<p>Возможные данные одного из&nbsp;файлов иллюстрированы графиком.</p>
<p><strong>Внимание! График приведён в&nbsp;иллюстративных целях для произвольных значений. Чтобы выполнить задание, используйте данные из&nbsp;файла в&nbsp;приложении.</strong></p>
<p><img src="https://avatars.mds.yandex.net/get-schoolbook-images/6887102/51fc6b3e22127b9242ce1e24cbb2a146/orig" alt="Screenshot"></p></div></div>

Ответ:

31 195

333 1520

In [67]:
b = []
with open("./3/27A.csv") as f:
    f.readline()
    for i in f.readlines():
        b.append(list(map(str, i.split(';'))))
        
        
st_year = set()
# 5 x 10
clust = [[[] for i in range(10)] for j in range(5)]
for i in b:
    v1, v2 = 0, 0
    ans = 0
    staz = int(i[6])
    ans = (int(i[4]) if i[4] != '' else int(i[5])) + (int(i[5]) if i[5] != '' else int(i[4]))

    st_year.add(ans // 100000)
    clust[staz // 2][ans // 100000].append(i)
    
for i in clust:
    for j in i:
        print(len(j), end=' ')
    print()

idx = [[1, 2], [2, 2]]
total = 111 + 113

py = 0

for i in idx:
    for j in clust[i[0]][i[1]]:
        # print(j, py)
        if "Python" in j[-1]:
            py += 1
print(total - py, py)

0 6 34 16 6 6 9 0 0 0 
0 6 111 51 24 20 5 0 2 0 
0 8 113 47 33 18 10 0 0 1 
0 2 38 16 5 7 2 1 0 1 
0 2 25 12 10 4 1 0 1 0 
194 30


In [66]:
b = []
with open("./3/27.csv") as f:
    f.readline()
    for i in f.readlines():
        b.append(list(map(str, i.split(';'))))
st_year = set()
# 5 x 10
clust = [[[] for i in range(10)] for j in range(5)]
for i in b:
    v1, v2 = 0, 0
    ans = 0
    staz = int(i[6])
    ans = (int(i[4]) if i[4] != '' else int(i[5])) + (int(i[5]) if i[5] != '' else int(i[4]))

    st_year.add(ans // 100000)
    clust[staz // 2][ans // 100000].append(i)

for i in clust:
    for j in i:
        print(len(j), end=' ')
    print()

idx = [[3, 2], [3, 1], [4, 1]]
total = 556 + 754 + 522 

py = 0

for i in idx:
    for j in clust[i[0]][i[1]]:
        # print(j)
        if "Python" in j[-1]:
            py += 1
print(total - py, py)

0 195 336 329 175 80 21 2 3 0 
0 165 316 299 166 76 23 1 0 1 
0 172 227 131 105 65 17 1 0 0 
0 556 754 417 313 181 96 0 1 0 
0 522 85 426 295 216 52 1 0 0 
1502 330


<div class="Markdown-module__Markdown--f5ALc TasksBankTaskCard-module__TaskCardContentMarkdown--hsXqP"><p>В лесу выделено два (файл А) или три (файл В) участка для вырубки деревьев. Размер каждого участка&nbsp;— 1&nbsp;гектар с&nbsp;<strong>целочисленными</strong> координатами границ. 1&nbsp;гектар = 10&nbsp;000 м². Деревья, не попавшие в&nbsp;выделенный гектар, спиливать нельзя. Участки для вырубки должны быть выбраны так, чтобы в&nbsp;них попадало наибольшее количество деревьев, при этом участки не могут пересекаться.</p>
<p>Известны координаты (X, Y) в метрах и высота каждого дерева (H). После спиливания дерево необходимо доставить до точки сбора — места для обработки и&nbsp;дальнейшей транспортировки. «Сложность» доставки дерева вычисляется как расстояние по&nbsp;Евклидовой метрике (<span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msqrt><mrow><mo stretchy="false">(</mo><msub><mi>x</mi><mn>1</mn></msub><mo>−</mo><msub><mi>x</mi><mn>2</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup><mo>+</mo><mo stretchy="false">(</mo><msub><mi>y</mi><mn>1</mn></msub><mo>−</mo><msub><mi>y</mi><mn>2</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></msqrt></mrow><annotation encoding="application/x-tex">\sqrt{(x_1-x_2)^2+(y_1-y_2)^2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1.24em; vertical-align: -0.305em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.935em;"><span class="svg-align" style="top: -3.2em;"><span class="pstrut" style="height: 3.2em;"></span><span class="mord" style="padding-left: 1em;"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: 0em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: 0em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height: 0.7401em;"><span style="top: -2.989em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: -0.0359em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: -0.0359em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height: 0.7401em;"><span style="top: -2.989em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span><span style="top: -2.895em;"><span class="pstrut" style="height: 3.2em;"></span><span class="hide-tail" style="min-width: 1.02em; height: 1.28em;"><svg xmlns="http://www.w3.org/2000/svg" width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.305em;"><span></span></span></span></span></span></span></span></span></span>) от&nbsp;его положения до точки сбора, умноженное на высоту дерева. Оптимальная точка сбора находится на месте одного из деревьев, попавших под вырубку. Выбор этой точки определяется минимальной суммарной сложностью доставки всех остальных деревьев вырубки до данного дерева. Нужно определить координаты двух точек сбора деревьев для файла&nbsp;А и&nbsp;трёх точек для файла&nbsp;В.</p>
<p><strong>Входные данные.</strong> Два файла, у которых в каждой строке хранятся три числа: координата по&nbsp;горизонтали (X), по&nbsp;вертикали&nbsp;(Y), высота дерева&nbsp;(H). Все величины указаны в&nbsp;метрах.</p>
<p><strong>Выходные данные.</strong> Для каждого файла найдите два числа: среднее арифметическое абсцисс (Px) и&nbsp;среднее арифметическое ординат (Py) точек сбора. В&nbsp;ответ запишите целую часть значений выражений <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>x</mi><mo>×</mo><mn>100000</mn></mrow><annotation encoding="application/x-tex">Px \times 100000</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.7667em; vertical-align: -0.0833em;"></span><span class="mord mathnormal" style="margin-right: 0.13889em;">P</span><span class="mord mathnormal">x</span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">×</span><span class="mspace" style="margin-right: 0.2222em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">100000</span></span></span></span></span> и <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>y</mi><mo>×</mo><mn>100000</mn></mrow><annotation encoding="application/x-tex">Py \times 100000</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.8778em; vertical-align: -0.1944em;"></span><span class="mord mathnormal" style="margin-right: 0.13889em;">P</span><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">×</span><span class="mspace" style="margin-right: 0.2222em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">100000</span></span></span></span></span>.</p>
<p><em>Пример.</em> Есть два участка вырубки, ограниченные размером 7&nbsp;на&nbsp;7&nbsp;метров.</p>
<p>Исходные данные:</p>
<pre><code>33.712    	14.021    	3
30.853    	13.394    	3
10.751    	17.077    	2
31.876    	09.725    	4
29.575    	13.998    	4
12.348    	19.989    	4
09.015    	22.504    	6
30.293    	12.751    	7
13.525    	22.514    	1
09.424    	17.026    	5
</code></pre>
<p>Построение графическое:<br>
<img src="https://avatars.mds.yandex.net/get-schoolbook-images/5545908/e309f88377413102233c9b0b806c07d8/orig" alt="Screenshot"></p>
<p>Визуально определяем границы участков.<br>
Первый — координаты всех точек укладываются в <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>x</mi><mo>∈</mo><mo stretchy="false">[</mo><mn>7</mn><mo separator="true">,</mo><mn>14</mn><mo stretchy="false">]</mo><mo separator="true">,</mo><mi>y</mi><mo>∈</mo><mo stretchy="false">[</mo><mn>16</mn><mo separator="true">,</mo><mn>23</mn><mo stretchy="false">]</mo><mi mathvariant="normal">.</mi></mrow><annotation encoding="application/x-tex">x \in [7,14], y \in [16,23].</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.5782em; vertical-align: -0.0391em;"></span><span class="mord mathnormal">x</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">∈</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mopen">[</span><span class="mord">7</span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord">14</span><span class="mclose">]</span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">∈</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mopen">[</span><span class="mord">16</span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord">23</span><span class="mclose">]</span><span class="mord">.</span></span></span></span></span><br>
Второй — <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>x</mi><mo>∈</mo><mo stretchy="false">[</mo><mn>28</mn><mo separator="true">,</mo><mn>35</mn><mo stretchy="false">]</mo><mo separator="true">,</mo><mi>y</mi><mo>∈</mo><mo stretchy="false">[</mo><mn>9</mn><mo separator="true">,</mo><mn>16</mn><mo stretchy="false">]</mo><mi mathvariant="normal">.</mi></mrow><annotation encoding="application/x-tex">x \in [28,35], y \in [9,16].</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.5782em; vertical-align: -0.0391em;"></span><span class="mord mathnormal">x</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">∈</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mopen">[</span><span class="mord">28</span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord">35</span><span class="mclose">]</span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">∈</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mopen">[</span><span class="mord">9</span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord">16</span><span class="mclose">]</span><span class="mord">.</span></span></span></span></span><br>
Определяем координаты центральных деревьев.<br>
Первый участок&nbsp;— (12.348, 19.989), второй&nbsp;— (30.293, 12.751).<br>
Ответ для примера: 2 132 050 и 1 637 000.</p></div>

Ответ:

25667971 24175730

-3333919 -46664

In [7]:
import math

a = []

with open("./4/27_A.txt") as f:
    f.readline()
    for i in f.readlines():
        a.append(list(map(float, i.replace(',', '.').split())))

clust = [[], []]

for i in a:
    # print(i)
    x, y, h = i
    clust_num = 0 if x < 150 and y < 250 else 1 if x > 350 and y > 250 else 2
    if clust_num == 2: continue
    clust[clust_num].append(i)


def dist(a, b):
    return math.sqrt((a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2) * b[2]

def sum_dist(a, clust):
    return sum(dist(a, j) for j in clust)

ans = []

for j in range(len(clust)):
    mini = 1e18
    v = -1
    for i in range(len(clust[j])):
        val = sum_dist(clust[j][i], clust[j])
        if val < mini:
            mini = val
            v = clust[j][i]
    ans.append(v)
    
print(sum(i[0] for i in ans) / 2 * 100000, sum(i[1] for i in ans) / 2 * 100000)

25667971.8549 24175730.56145


<div class="Text Text_typography_bodyM TasksBankTaskCard-module__TaskCardContent--PNxia"><div class="Markdown-module__Markdown--f5ALc TasksBankTaskCard-module__TaskCardContentMarkdown--hsXqP"><p>Учёный решил провести кластеризацию некоторого множества звёзд по&nbsp;их&nbsp;расположению на&nbsp;карте звёздного неба. Кластер звёзд&nbsp;— это набор звёзд (точек) на&nbsp;графике, лежащий внутри прямоугольника высотой <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>H</mi></mrow><annotation encoding="application/x-tex">H</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.6833em;"></span><span class="mord mathnormal" style="margin-right: 0.08125em;">H</span></span></span></span></span> и&nbsp;шириной <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>W</mi></mrow><annotation encoding="application/x-tex">W</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.6833em;"></span><span class="mord mathnormal" style="margin-right: 0.13889em;">W</span></span></span></span></span>. Каждая звезда обязательно принадлежит только одному из&nbsp;кластеров.</p>
<p>Истинный центр кластера, или <strong>центроид</strong>,&nbsp;— это одна из&nbsp;звёзд на&nbsp;графике, сумма расстояний от&nbsp;которой до&nbsp;всех остальных звёзд кластера минимальна.</p>
<p>Под расстоянием понимается расстояние Евклида между двумя точками <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi><mo stretchy="false">(</mo><msub><mi>x</mi><mn>1</mn></msub><mo separator="true">,</mo><msub><mi>y</mi><mn>1</mn></msub><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">A (x_1, y_1)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mord mathnormal">A</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: 0em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: -0.0359em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mclose">)</span></span></span></span></span> и&nbsp;<span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>B</mi><mo stretchy="false">(</mo><msub><mi>x</mi><mn>2</mn></msub><mo separator="true">,</mo><msub><mi>y</mi><mn>2</mn></msub><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">B (x_2, y_2)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mord mathnormal" style="margin-right: 0.05017em;">B</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: 0em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: -0.0359em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mclose">)</span></span></span></span></span> на&nbsp;плоскости, которое вычисляется по&nbsp;формуле:</p>
<p><span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>d</mi><mo stretchy="false">(</mo><mi>A</mi><mo separator="true">,</mo><mi>B</mi><mo stretchy="false">)</mo><mo>=</mo><msqrt><mrow><mo stretchy="false">(</mo><msub><mi>x</mi><mn>2</mn></msub><mtext>&nbsp;—</mtext><msub><mi>x</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup><mo>+</mo><mo stretchy="false">(</mo><msub><mi>y</mi><mn>2</mn></msub><mtext>&nbsp;—</mtext><msub><mi>y</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></msqrt><mi mathvariant="normal">.</mi></mrow><annotation encoding="application/x-tex">d (A, B) = \sqrt{(x_2&nbsp;— x_1)^2 + (y_2&nbsp;— y_1)^2}.</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord mathnormal">A</span><span class="mpunct">,</span><span class="mspace" style="margin-right: 0.1667em;"></span><span class="mord mathnormal" style="margin-right: 0.05017em;">B</span><span class="mclose">)</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 1.24em; vertical-align: -0.305em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.935em;"><span class="svg-align" style="top: -3.2em;"><span class="pstrut" style="height: 3.2em;"></span><span class="mord" style="padding-left: 1em;"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: 0em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mord">&nbsp;—</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: 0em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height: 0.7401em;"><span style="top: -2.989em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: -0.0359em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mord">&nbsp;—</span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span style="top: -2.55em; margin-left: -0.0359em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height: 0.7401em;"><span style="top: -2.989em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span><span style="top: -2.895em;"><span class="pstrut" style="height: 3.2em;"></span><span class="hide-tail" style="min-width: 1.02em; height: 1.28em;"><svg xmlns="http://www.w3.org/2000/svg" width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.305em;"><span></span></span></span></span></span><span class="mord">.</span></span></span></span></span></p>
<p>В&nbsp;файле A&nbsp;хранятся данные о&nbsp;звёздах <strong>двух</strong> кластеров, где <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>H</mi><mo>=</mo><mn>3</mn></mrow><annotation encoding="application/x-tex">H=3</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.6833em;"></span><span class="mord mathnormal" style="margin-right: 0.08125em;">H</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">3</span></span></span></span></span>, <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>W</mi><mo>=</mo><mn>3</mn></mrow><annotation encoding="application/x-tex">W=3</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.6833em;"></span><span class="mord mathnormal" style="margin-right: 0.13889em;">W</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">3</span></span></span></span></span> для каждого кластера. В&nbsp;каждой строке записана информация о&nbsp;расположении на&nbsp;карте одной звезды: сначала координата <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>x</mi></mrow><annotation encoding="application/x-tex">x</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.4306em;"></span><span class="mord mathnormal">x</span></span></span></span></span>, затем координата <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>y</mi></mrow><annotation encoding="application/x-tex">y</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.625em; vertical-align: -0.1944em;"></span><span class="mord mathnormal" style="margin-right: 0.03588em;">y</span></span></span></span></span>. Значения даны в&nbsp;условных единицах. Известно, что количество звёзд не&nbsp;превышает 1000.</p>
<p>В&nbsp;файле Б&nbsp;хранятся данные о&nbsp;звёздах <strong>трёх</strong> кластеров, где <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>H</mi><mo>=</mo><mn>3</mn></mrow><annotation encoding="application/x-tex">H=3</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.6833em;"></span><span class="mord mathnormal" style="margin-right: 0.08125em;">H</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">3</span></span></span></span></span>, <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>W</mi><mo>=</mo><mn>3</mn></mrow><annotation encoding="application/x-tex">W=3</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.6833em;"></span><span class="mord mathnormal" style="margin-right: 0.13889em;">W</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">3</span></span></span></span></span> для каждого кластера. Известно, что количество звёзд не&nbsp;превышает 10&nbsp;000. Структура хранения информации о&nbsp;звездах в&nbsp;файле Б&nbsp;аналогична файлу&nbsp;А.</p>
<p>Для каждого файла определите координаты центра каждого кластера, затем вычислите два числа: <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>P</mi><mi>x</mi></msub></mrow><annotation encoding="application/x-tex">P_x</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.8333em; vertical-align: -0.15em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.1514em;"><span style="top: -2.55em; margin-left: -0.1389em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">x</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span></span></span></span></span>&nbsp;— среднее арифметическое абсцисс центров кластеров, и&nbsp;<span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>P</mi><mi>y</mi></msub></mrow><annotation encoding="application/x-tex">P_y</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.9694em; vertical-align: -0.2861em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.1514em;"><span style="top: -2.55em; margin-left: -0.1389em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right: 0.03588em;">y</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.2861em;"><span></span></span></span></span></span></span></span></span></span></span>&nbsp;— среднее арифметическое ординат центров кластеров.</p>
<p>В&nbsp;ответе запишите четыре числа: в&nbsp;первой строке сначала целую часть произведения <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>P</mi><mi>x</mi></msub><mo>×</mo><mn>10</mn><mtext>&nbsp;</mtext><mn>000</mn></mrow><annotation encoding="application/x-tex">P_x \times 10&nbsp;000</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.8333em; vertical-align: -0.15em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.1514em;"><span style="top: -2.55em; margin-left: -0.1389em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">x</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">×</span><span class="mspace" style="margin-right: 0.2222em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">10&nbsp;000</span></span></span></span></span>, затем целую часть произведения <span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>P</mi><mi>y</mi></msub><mo>×</mo><mn>10</mn><mtext>&nbsp;</mtext><mn>000</mn></mrow><annotation encoding="application/x-tex">P_y \times 10&nbsp;000</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.9694em; vertical-align: -0.2861em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right: 0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.1514em;"><span style="top: -2.55em; margin-left: -0.1389em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right: 0.03588em;">y</span></span></span></span><span class="vlist-s">&ZeroWidthSpace;</span></span><span class="vlist-r"><span class="vlist" style="height: 0.2861em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">×</span><span class="mspace" style="margin-right: 0.2222em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">10&nbsp;000</span></span></span></span></span> для файла A, во&nbsp;второй строке&nbsp;— аналогичные данные для файла Б.</p>
<p>Возможные данные одного из&nbsp;файлов иллюстрированы графиком.</p>
<p><strong>Внимание! График приведён в&nbsp;иллюстративных целях для произвольных значений, не&nbsp;имеющих отношения к&nbsp;заданию.</strong></p>
<p><strong>Для выполнения задания используйте данные из&nbsp;прилагаемого файла.</strong></p>
<p><img src="https://avatars.mds.yandex.net/get-schoolbook-images/5519467/4013b2d69bb671fbe807dbedb7e82ccc/orig" alt="Screenshot"></p></div></div>

Ответ:

10738 30730

37522 51277

In [1]:
import math

a = []

with open("./5/27_A (1).txt") as f:
    f.readline()
    for i in f.readlines():
        a.append(list(map(float, i.replace(',', '.').split())))

clust = [[], []]

for i in a:
    if i[0] > 1:
        clust[1].append(i)
    else:
        clust[0].append(i)

def dist(a, b):
    return math.sqrt((a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2)

def sum_dist(a, clust):
    return sum(dist(a, j) for j in clust)

ans = []


mini = 1e9
idx_0 = -1
for i in range(len(clust[0])):
    val = sum_dist(clust[0][i], clust[0])
    if val < mini:
        mini = val
        idx_0 = i
# print(clust[0][idx_0])


mini = 1e9
idx_1 = -1
for i in range(len(clust[1])):
    val = sum_dist(clust[1][i], clust[1])
    if val < mini:
        mini = val
        idx_1 = i
# print(clust[1][idx_1])

print((clust[0][idx_0][0] + clust[1][idx_1][0]) / 2, (clust[0][idx_0][1] + clust[1][idx_1][1]) / 2)


1.073821226546789 3.07300760590781


In [2]:
import math

a = []

with open("./5/27_B (1).txt") as f:
    f.readline()
    for i in f.readlines():
        a.append(list(map(float, i.replace(',', '.').split())))

clust = [[], [], []]

for i in a:
    clust_num = 2 if i[0] > 5 else 1 if i[1] > 7.2 else 0
    clust[clust_num].append(i)

def dist(a, b):
    return math.sqrt((a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2)

def sum_dist(a, clust):
    return sum(dist(a, j) for j in clust)

ans = []

for j in range(len(clust)):
    mini = 1e9
    v = -1
    for i in range(len(clust[j])):
        val = sum_dist(clust[j][i], clust[j])
        if val < mini:
            mini = val
            v = clust[j][i]
    ans.append(v)
    
print(sum(i[0] for i in ans) / 3, sum(i[1] for i in ans) / 3)


3.7522944615707168 5.127795880214987
